File: aoc-2020-kotlin/src/main/kotlin/days/day06/day06.kt
package days.day06

import lib.Reader

fun main() {
    val input = Reader("day06.txt").string().split("\n\n")

    val part1 = input.sumBy { it.replace("\n", "").toSet().size }

    println(part1)

    val part2 = input
            .map { it.split("\n") }
            .sumBy { group ->
                ('a'..'z').count { letter ->
                    group.all { it.contains(letter) }
                }
            }

    println(part2)

    val part1again = input
        .sumBy {
            it.split("\n")
                .map(String::toSet)
                .reduce { acc, list -> acc.union(list) }
                .size
        }

    println(part1again)

    val part2again = input
        .sumBy {
            it.split("\n")
                .map(String::toSet)
                .reduce { acc, list -> acc.intersect(list) }
                .size
        }

    println(part2again)
}

File: aoc-2016-clojure/src/aoc_2016_clojure/day06.clj
(ns aoc-2016-clojure.day06
  (:require [clojure.string :as str]
            [clojure.java.io :as io]))

(def testInput (str/split "eedadn\ndrvtee\neandsr\nraavrd\natevrs\ntsrnev\nsdttsa\nrasrtv\nnssdts\nntnada\nsvetve\ntesnvt\nvntsnd\nvrdear\ndvrsen\nenarar" #"\n"))
(def puzzleInput (str/split (slurp (io/resource "day06.txt")) #"\n"))

(defn increase-letter-count [[letter counts]]
  (let [letter-key (keyword (str letter))]
    (update counts letter-key #(if (not %) 1 (inc %)))))

(defn increase-letters-counts [counts word]
  (let [zipped (map vector word counts)]
    (map increase-letter-count zipped)))

(defn count-letter-frequencies [words]
  (let [empty-counts (take (count (first words)) (repeat {}))]
    (reduce increase-letters-counts empty-counts words)))

(defn get-key-by-comparison [comparison input]
  (reduce
    (fn [acc next]
      (if (comparison (second acc) (second next))
        acc
        next))
    input))

(defn decoder [input ordering]
  (let [frequencies   (count-letter-frequencies input)
        most-frequent (map (partial get-key-by-comparison ordering) frequencies)]
    (str/join (map (comp name first) most-frequent))))

(defn part1 [input] (decoder input >))
(defn part2 [input] (decoder input <))

(part1 testInput)
(part2 testInput)
(part1 puzzleInput)
(part2 puzzleInput)
